### PR TITLE
#272 migrate migration scripts

### LINKED ISSUE #272: Migrate migration scripts
## Expected Behaviour

- Migration/import scripts stored in the `unified-codes` repo
- Scripts adjusted to cater to updated (v2?) schema

## Current Behaviour

- Scripts are stored in a separate repo (`universal-drug-codes-server`) and cater to V1 of the unified codes data schema

## Possible Solution

- Place scripts somewhere like `unified-codes/tools/scripts/dgraph/migrate`
- Consolidate/tidy up while migrating? 🤷 

NB: The actual scripts are pending some confirmation on expected spreadsheet format!

### PR DESCRIPTION
**Pull request by katherine-sussol**
Partially Fixes #272 and #266.

## Description

v2 of the import script (imports products/codes from excel spreadsheet into Dgraph) - still pending on finalised spreadsheet data.

**v1**
- Added v1 scripts to `unified-codes` repo (thanks @mark-prins - stolen from `universal-drug-codes-server`)

**v2 - Imports via Dgraph mutations (using Dgraph's GRPC client)**
- Creates root category nodes (Drug, Consumable, Other)
- Imports products from excel spreadsheet under the appropriate category (ignoring combinations for now) using upserts (will create node if it doesn't exist, otherwise it will modify existing one)
- Adds synonyms to product

Have added a sample Dgraph schema for testing (as the schema needs to be defined before import), I think once #266 is finalised can remove and/or rename any types required.

Next steps: 
- Wait for spreadsheet finalisation and apply any necessary tweaks.

**To Test:**
- In `tools/scripts/dgraph` run:
  - `bash ./clean.sh` 
  - `bash ./alter.sh ./data/v2/schema.gql` 
- In `tools/scripts/dgraph/migration/v2` run:
  - `node import.js`
- In Ratel enter the following query to view results:
```
{
    query(func: eq(dgraph.type, Category)) @recurse(loop:false) {
      uid
      code
      dgraph.type
      name@*
      type
      value
      combines
      children
      properties
    }
}
```
 
### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [ ] Documentation OK.

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [ ] Tests OK.

### CODE REVIEW

**Comment by katherine-sussol** [comment]
Converted to draft as there are ongoing format revisions - will continue to add commits here, feedback welcome while in progress.

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 1]
```diff
@@ -0,0 +1,70 @@
+type Property {
+  properties
```
Comment (mark-prins) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:2]: is this so that a property can have child properties?

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+};
+
+const insertProduct = async (productName, productCode, categoryCode, synonymsArray) => {
+  const query = `query {
+    Category as var(func: eq(dgraph.type, Category)) @filter(eq(code, ${categoryCode}))
+    Product as var(func: eq(code, ${productCode})) 
+  }`;
+
+  let synonymMutation = '';
+  for (let i = 0; i < synonymsArray.length; i++) {
```
Comment (mark-prins) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:62]: perfect as it is. just for fun, also possible using `reduce()`

```
const synonymMutation = synonymsArray.reduce((mutation, synonym, i) => `${mutation}uid(Product) <name@alt${i}> "${synonym.trim()}" .\n`, '');

```

**Review by mark-prins — COMMENTED** [review]
👍  from me! looks great :)

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js (hunk 2 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/migration/v2/import.js:62]: Ooh fancy 👍  thanks for that

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 1 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:2]: Oh, yes it was for this reason - based on the old schema. But recalling our previous discussion we probably don't need it anymore, will remove.

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 3]
```diff
@@ -0,0 +1,83 @@
+type Property {
+  type
+  value
+}
+
+type Category {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:6]:
```
export interface ICategoryEntity extends IEntity {
  name: ECategoryEntityName;
  type: EEntityType.Category;
  children?: ICategoryEntityChildren;
  properties?: ICategoryEntityProperties;
}
```

yes_1

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  children
+  properties
+}
+
+type Product {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:13]:
```
export interface IProductEntity extends IEntity {
  type: EEntityType.Product;
  combines?: IProductEntity[];
  children?: IProductEntityChildren;
  properties?: IProductEntityProperties;
}
```

yes_2

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  combines
+  children
+  properties
+}
+
+type Route {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:21]:
```
export interface IRouteEntity extends IEntity {
  type: EEntityType.Route;
  children?: IRouteEntityChildren;
  properties?: IRouteEntityProperties;
}
```

yes_3

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  children
+  properties
+}
+
+type DoseForm {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:28]:
```
export interface IFormEntity extends IEntity {
  type: EEntityType.Form;
  children?: IFormEntityChildren;
  properties?: IFormEntityProperties;
}
```

Type is currently just `FormEntity`, suggest changing it to `DoseFormEntity`?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  children
+  properties
+}
+
+type DoseQualification {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:35]:
```
export interface IFormQualifierEntity extends IEntity {
  code: string;
  name: string;
  type: EEntityType.FormQualifier;
  children?: IFormQualifierEntityChildren;
  properties?: IFormQualifierEntityProperties;
}
```

Type is `FormQualifier`, maybe meet in the middle with `DoseFormQualifier`?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  children
+  properties
+}
+
+type DoseUnit {
+  code
+  name
+  children
+  properties
+}
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:54]:
```
```

export interface IUnitEntity extends IEntity {
  type: EEntityType.Unit;
  children?: IUnitEntityChildren
  properties?: IUnitEntityProperties;
}
```

Oops, this is more up-to-date than the types, will need to split `UnitEntity` into `DoseStrengthEntity` and `DoseUnitEntity`?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+type InnerPack {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:56]:
```
export interface IPackImmediateEntity extends IEntity {
  code: string;
  name: string;
  type: EEntityType.PackImmediate;
  children?: IPackImmediateEntityChild[];
  properties?: IPackImmediateEntityProperty[];
}
```

Hmm, type is currently `PackImmediateEntity`, do we want to go with `InnerPack` or `PackImmediate` (or some other variation of)?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  children
+  properties
+}
+
+type PackSize {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:63]:
```
export interface IPackSizeEntity extends IEntity {
  code: string;
  name: string;
  type: EEntityType.PackSize;
  children?: IPackSizeEntityChild[];
  properties?: IPackSizeEntityProperty[];
}
```

yes_4

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  children
+  properties
+}
+
+type OuterPack {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:70]:
```
export interface IPackOuterEntity extends IEntity {
  code: string;
  name: string;
  type: EEntityType.PackOuter;
  children?: IPackOuterEntityChild[];
  properties?: IPackOuterEntityProperty[];
}
```

Hmm, `PackOuter` or `OuterPack`?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  code
+  name
+  children
+  properties
+}
+
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/data/v2/schema.gql:76]: Do we want to add `Manufacturer` and `Brand`, or can that wait?

**Review by wlthomson — COMMENTED** [review]
First pass, haven't tested code yet, just comments on the `v2` gql schema.

This covers quite a lot of https://github.com/openmsupply/unified-codes/issues/266 (issue stealer!), so any significant changes can be addressed in that PR, I think?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 6 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:28]: Yep yep, can do!

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 12 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:76]: Good question, none of our current spreadsheet data has any Manufacturer or Brand details so I thought maybe best to revisit when they do start showing up?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 7 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:35]: Works for me!

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 9 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:56]: I think in this instance `PackImmediate` is probably better than `InnerPack`.
It's called `Immediate Packaging` in the spreadsheet but that's a bit long...

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 11 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:70]: I think `OuterPack` sounds better? And also matches the terminology Rowena and Emily are using more closely (which is `Outer Packaging`)

File: unified-codes/tools/scripts/dgraph/migration/v1/export.js [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+    case 'form':
+      parent.form = description;
+      parseChildren(item.has_child, parent);
+      break;
+    case 'form_category':
+      parent.formCategory = description;
+      parseChildren(item.has_child, parent);
+      break;
+  }
+};
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/export.js:83]: Don't want to mess with the recursion here too much, but I find it's a bit easier when recursive functions just call themselves with parameters, e.g. instead of `parseProduct` and `parseChildren` can have a generic `parseEntity` which handles both cases (can do a `forEach` for the children, and can look at the properties of the `entity` passed as a param to see how to handle, maybe with a `case` statement or some `if` statements)?

File: unified-codes/tools/scripts/dgraph/migration/v1/export.js [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+  const formB = itemB.form;
+  if (formA < formB) {
+    return -1;
+  }
+  if (formA > formB) {
+    return 1;
+  }
+
+  return 0;
+};
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/export.js:114]: Hmm, does this work as expected? I've had some weird behaviour with multi-case sorts like this. Since the sort will use some sort of sublinear algorithm, it won't compare every item with every item (in mathsy terms, the sort function needs to be distributive, which I think explodes with multi-case comparisons). I'd expect this to work mostly, but for some items to get thrown in in random places... but if it's working, I say leave it 🤷‍♂️ .

File: unified-codes/tools/scripts/dgraph/migration/v1/export.js [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+    row.getCell(cell++).value = item.form;
+    row.getCell(cell++).value = item.code;
+    row.getCell(cell++).value = item.description;
+    row.getCell(cell++).value = item.strength;
+    row.getCell(cell++).value = item.ddd;
+    row.getCell(cell++).value = item.rxnav;
+    row.getCell(cell++).value = item.eml;
+    row.getCell(cell++).value = item.atc;
+    row.getCell(cell++).value = item.unspsc;
+    row.getCell(cell++).value = item.nzulm;
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/export.js:166]: If you made an array of, e.g. `["code", "description", ...]`, you could prob do this with a `forEach` function, e.g.

```
itemProperties.forEach((property) => row.getCell(cell++).value = item[property]);
```

Might be a bit more tricky with the `product.x` bits, but could nest that bit?

File: unified-codes/tools/scripts/dgraph/migration/v1/import.js [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+  'code',
+  'title',
+  'strength',
+  'ddd',
+  'rxnav',
+  'eml',
+  'atc',
+  'unspsc',
+  'nzulm',
+];
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/import.js:30]: See previous comment... something like this!

yes_2

File: unified-codes/tools/scripts/dgraph/migration/v1/import.js [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+const addPropertiesToProduct = (product, colNumber, value) => {
+  const property = properties[colNumber - 1];
+  product[property] = value;
+};
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/import.js:34]: Hmm... I think this is only used in `processRow`, could maybe be a nested function?

File: unified-codes/tools/scripts/dgraph/migration/v1/import.js [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+
+const convertItemToProduct = (item) => {
+  const { code, title: description, type, has_child } = item;
+  const product = {
+    code,
+    description,
+    has_child,
+    has_property: [],
+    type,
+  };
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/import.js:44]: Nit: minor tidy up?

```suggestion
  const { code, title: description, type, has_child = []. has_property = [] } = item;
  const product = {
    code,
    description,
    has_child,
    has_property,
    type,
  };
```

File: unified-codes/tools/scripts/dgraph/migration/v1/import.js [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+  //if (item.type) product.type = item.type;
+  if (item.strength) {
+    product.has_property.push({ value: item.strength, type: 'strength' });
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/import.js:47]: As above, maybe some constants for property types?

File: unified-codes/tools/scripts/dgraph/migration/v1/import.js [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    form = {
+      description: item.form,
+      type: 'form',
+      has_child: [],
+    };
+    formCategory.has_child.push(form);
+  }
+
+  form.has_child.push(convertItemToProduct(item));
+};
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v1/import.js:132]: Hmm, I think some of this stuff will be good to extract into a class (maybe a follow up PR). It'll be a bit easier to "bump" as changes get made if we use a more OOP design, I think?

Priority at this stage is getting it working, and this works, so... 👍 !

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+  mutation.setSetNquads(`
+    uid(DrugCategory) <name> "Drug" .
+    uid(ConsumableCategory) <name> "Consumable" .
+    uid(OtherCategory) <name> "Other" .
+    uid(DrugCategory) <dgraph.type> "Category" .
+    uid(ConsumableCategory) <dgraph.type> "Category" .
+    uid(OtherCategory) <dgraph.type> "Category" .
+    uid(DrugCategory) <code> "933f3f00" .
+    uid(ConsumableCategory) <code> "77fcbb00" .
+  `);
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:28]: Nquads! Nice!

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+  const atcUC5 = row.getCell(27).value;
+  const dddUC5 = row.getCell(28).value;
+  const emlUC2 = row.getCell(29).value;
+  const emlUC67 = row.getCell(30).value;
+  const unspsc = row.getCell(31).value;
+  const nzulmUC2 = row.getCell(32).value;
+  const nzulmUC67 = row.getCell(33).value;
+  const snomed = row.getCell(34).value;
+  const drugbank = row.getCell(35).value;
+  const usfda = row.getCell(36).value;
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:73]: I think could probably make this a bit fancy with some functional stuff, e.g.

```
propertyCells = {
  gs1: 22,
  rvNavUC2: 23,
  ...
};

const properties = Object.values(cellProperties).reduce( .... );
```

Would then end up with an object like:

```
{
  gs1: "12345678",
  rxNavUC2: "987654321",
  ...
}
```

where `"123456789"` etc. are the values taken from the spreadsheet.

Again, not urgent... something which can be refined later!

smile

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+  let parentIndex = 0;
+  let childIndex = 1;
+  while (childIndex < productDefinition.length) {
+    if (productDefinition[childIndex].name && productDefinition[childIndex].code) {
+      await insertChild(productDefinition[parentIndex], productDefinition[childIndex]);
+      parentIndex = childIndex;
+    }
+    childIndex++;
+  }
+};
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:146]: A little bit hard to follow... but works! Again, might benefit from being extracted out into a dedicated class method, just to make it a bit more modular.

As usual, not a priority atm!

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+  // const propertiesMutations = propertyMutationsForNode(product);
+
+  const query = `query {
+    Category as var(func: eq(dgraph.type, Category)) @filter(eq(code, ${categoryCode}))
+    ${product.type} as var(func: eq(code, ${product.code})) 
+  }`;
+
+  const synonymMutation = synonymsArray.reduce(
+    (mutation, synonym, i) =>
+      `${mutation}uid(${product.type}) <name@alt${i}> "${synonym.trim()}" .\n`,
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:159]: 👍 .

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+    ''
+  );
+  const mutation = new dgraph.Mutation();
+  mutation.setSetNquads(`
+    uid(${product.type}) <name> "${product.name}" .
+    uid(${product.type}) <code> "${product.code}" .
+    uid(${product.type}) <dgraph.type> "${product.type}" .
+    uid(Category) <children> uid(${product.type}) .
+    ${synonymMutation}
+  `);
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:169]: Again, same thing, same thing (sorry 😢 ). Might be good to have queries as class methods/static fields, just for easier maintenance and stuff?

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 26]
```diff
@@ -0,0 +1,294 @@
...
+          ProductCode as var(func: eq(code, ${productCode}))
+        }`;
+
+        const mutation = new dgraph.Mutation();
+        mutation.setSetNquads(`
+          uid(Component) <combines> uid(ProductCode) .
+        `);
+
+        await processDgraphUpsert(query, mutation);
+      }
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:249]: Could prob get away with a `reduce` or `forEach` here... but just a style thing!

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+  node.properties.reduce(
+    (mutation, property, i) =>
+      property.value
+        ? `${mutation}uid(propUid${i}) <type> "${property.type}" .
+      uid(propUid${i}) <value> "${property.value}" .
+      uid(propUid${i}) <dgraph.type> "Property" .
+      uid(${node.type}) <properties> uid(propUid${i}) .\n`
+        : ``,
+    ''
+  );
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:291]: ...but what about a class!? 

laugh

**Review by wlthomson — APPROVED** [review]
Phew... made it!

lgtm

TLDR: 

- Works!
- A script, so doesn't need to be too fancy... but a fair bit of baked in magic numbers and constants dependent on the spreadsheet, so might be a bit hard to maintain.
- So... when time permits, maybe migrate a bit of the logic to some classes 😨 ?

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js [hunk 28]
```diff
@@ -0,0 +1,294 @@
...
+      type: 'DoseUnit',
+      name: doseUnitName,
+      code: doseUnitCode,
+      properties: [],
+    },
+  ];
+
+  const [product] = productDefinition;
+
+  if (product.name.indexOf('/') > -1) {
```
Comment (wlthomson) [inline unified-codes/tools/scripts/dgraph/migration/v2/import.js:126]: Think I'm getting an error here:

```
TypeError: Cannot read property 'indexOf' of null
    at processRow (C:\Users\willt\Projects\sussol\unified-codes\unified-codes\tools\scripts\dgraph\migration\v2\import.js:136:22)
    at createImport (C:\Users\willt\Projects\sussol\unified-codes\unified-codes\tools\scripts\dgraph\migration\v2\import.js:239:13)
```

Looks like `productName` is `null`? Could be a spreadsheet issue?

**Review by wlthomson — CHANGES_REQUESTED** [review]
Oops, found an error! 

surprise

Might be data related, rather than a script issue, but will request changes just to be safe!

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 12 above)
Comment (mark-prins) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:76]: I thought that brand / manufacturer could wait - we don't have any information for them yet. Will need to support GTINs at some point, but there's just no data yet.

File: unified-codes/tools/scripts/dgraph/migration/v1/add_codes.js [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+const outputFilename = process.argv[3] || './data/data.json';
+
+let matched = 0;
+
+const mutateItem = (item) => {
+  if (item.code) return item;
+  item.code = parser.generateCode();
+  matched++;
+  return item;
+};
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply unified-codes/tools/scripts/dgraph/migration/v1/add_codes.js:15]: mea culpa - and yes, I felt dirty while doing it. I tried functional and gave up in the end, and went quick and dirty.

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 8 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:54]: Yeah I think so! I had a question for Rowena on this one actually though as it seems the current spreadsheet has some missing codes for strength and/or unit although there is text defined.

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+code: string @index(exact, fulltext).
+name: string @lang @index(exact, term, trigram) .
+type: string @index(term) .
```
[the remark this answers is not in the document]
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:79]: The type defined here is mainly for the `Property` Dgraph type (which currently only has `type` and `value` properties). I agree bit confusing to add to all the other types. Seems a bit redundant?

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js (hunk 28 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/migration/v2/import.js:126]: I'm getting the same here at the end of the spreadsheet - I think because there are still technically rows in the worksheet, they are just blank. Can probably tidy it up to terminate more tidily when a blank product name is encountered.

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js (hunk 27 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/migration/v2/import.js:291]: I guess I just have no class... ...

An excellent suggestion nevertheless.

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 9 above)
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:56]: Yeah I say go with `PackImmediate`!

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 11 above)
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:70]: Hmm, just thinking for consistency with `PackImmediate` and `PackSize`. `PackOuter` follows the pattern, I guess?

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 12 above)
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:76]: OK, concensus seems to be no `Manufacturer` or `Brand` needed!

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 30 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:79]: Hmm, you're right actually. We will need to add to the `gql` mappers, but I don't think we need if for Dgraph at all (think we can get away with it for properties too, since we will have instances for each one)!

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js (hunk 27 above)
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/migration/v2/import.js:291]: > I guess I just have no class... ...

facepalm

> An excellent suggestion nevertheless. Perhaps we could add a tidy up issue to the backlog while we plough ahead with finishing up v1 (v1 of V2 UC?)

Sounds good!

File: unified-codes/tools/scripts/dgraph/migration/v2/import.js (hunk 28 above)
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/migration/v2/import.js:126]: Yeah, it's no biggie, everything still works, so no need to prioritise here, I don't think?

**Comment by wlthomson** [comment]
@katherine-sussol I'm happy for this to go in now, then we can get moving on the `v1` -> `v2` resolvers!

gogo

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 8 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql:54]: Just a small update on this, sounds like the missing codes are just a case of the spreadsheet being still in progress (which should be ready end of this week). Still valid structure woo 🥳

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 7 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql]: edcf0cef8

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 9 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql]: 2a792f9cf

File: unified-codes/tools/scripts/dgraph/data/v2/schema.gql (hunk 11 above)
Comment (katherine-sussol) [reply unified-codes/tools/scripts/dgraph/data/v2/schema.gql]: Yeah, the consistency is nice. Updated with 5d60a3499 (we currently don't have any codes for this level so just a schema update)

**Comment by katherine-sussol** [comment]
> @katherine-sussol I'm happy for this to go in now, then we can get moving on the `v1` -> `v2` resolvers!
> 
> gogo

Cool sounds good to me. I've renamed the levels that we agreed on and will create separate issue for tidy up later 🥳 .
I think as spreadsheet updates come in we can just create new issues for them (ETA from Rowena on next version is end of this week).
